### PR TITLE
Modify `CoffeeRemoteList` to store the git HEAD 

### LINKED ISSUE #186: Store the git HEAD information on the repository
We should store and display the git HEAD for each repo and print it in a table column in the terminal.

Maybe also the date of the last commit would be useful to display too

### PR DESCRIPTION
**Pull request by tareknaser**
Fixes #186

### CODE REVIEW

**Comment by vincenzopalazzo** [comment]
The solution is inside the error

```
coffee_1  |    Compiling tests v0.1.0 (/workdir/tests)
coffee_1  | error[E0283]: type annotations needed
coffee_1  |   --> tests/src/logger.rs:42:21
coffee_1  |    |
coffee_1  | 42 |                     DateTime::from(SystemTime::now())
coffee_1  |    |                     ^^^^^^^^^^^^^^ cannot infer type for struct `chrono::DateTime<_>`
coffee_1  |    |
coffee_1  |    = note: multiple `impl`s satisfying `chrono::DateTime<_>: From<SystemTime>` found in the `chrono` crate:
coffee_1  |            - impl From<SystemTime> for chrono::DateTime<Local>;
coffee_1  |            - impl From<SystemTime> for chrono::DateTime<Utc>;
coffee_1  | 
coffee_1  | For more information about this error, try `rustc --explain E0283`.
coffee_1  | error: could not compile `tests` (lib test) due to previous error
coffee_1  | make: *** [Makefile:35: integration] Error 101
```

But this depends from your code, so I think you should just play a little bit of what type `SystemTime::now()` return

File: coffee_cmd/src/coffee_term/command_show.rs [hunk 1]
```diff
@@ -54,6 +54,8 @@ pub fn show_remote_list(remote_list: Result<CoffeeRemote, CoffeeError>) -> Resul
         term::format::bold(String::from("Repository Alias")),
         term::format::bold(String::from("URL")),
         term::format::bold(String::from("N. Plugins")),
+        term::format::bold(String::from("Git HEAD")),
+        term::format::bold(String::from("Date of last commit")),
```
Comment (vincenzopalazzo) [inline coffee_cmd/src/coffee_term/command_show.rs:58]: s/"Date of last commit"/"last update"

It is better `Date of last commit`

File: coffee_core/src/coffee.rs [hunk 2]
```diff
@@ -396,10 +396,20 @@ impl PluginManager for CoffeeManager {
...
         let mut remote_list = Vec::new();
         for repo in self.repos.values() {
+            let commit_id = match repo.git_head() {
+                Some(val) => val,
+                None => String::from(""),
+            };
+            let date = match repo.date_of_the_last_commit() {
+                Some(val) => val,
+                None => String::from(""),
+            };
```
Comment (vincenzopalazzo) [inline coffee_core/src/coffee.rs:406]: Why do not store the date as a string if we have to store something, or if it is None we keep it None?

In rust there is the optional type so we should use it

File: coffee_github/Cargo.toml [hunk 3]
```diff
@@ -13,3 +13,4 @@ log = "0.4.17"
 env_logger = "0.9.3"
 serde_yaml = "^0.9.0"
 walkdir = "2.3.2"
+chrono = { version = "0.4", features = ["std"], default-features = false }
```
Comment (vincenzopalazzo) [inline coffee_github/Cargo.toml:16]: try to use this https://crates.io/crates/localtime/1.3.0

File: coffee_lib/src/repository.rs [hunk 4]
```diff
@@ -33,5 +33,14 @@ pub trait Repository: Any {
...
 
+    /// return the git head of the repository.
+    fn git_head(&self) -> Option<String>;
+
+    /// return the date of the last commit of the repository.
+    fn date_of_the_last_commit(&self) -> Option<String>;
+
+    /// return the git information of the repository.
+    async fn git_info(&self) -> Result<(String, String), CoffeeError>;
+
```
Comment (vincenzopalazzo) [inline coffee_lib/src/repository.rs]: If you noted inside the repository we do not have any reference to git, so the coffee_lib specifies only the central concept of the repository, but the repository is not coupled with git at all.

The git implementation is the coffee_github one, and where these methods should live too

File: coffee_github/src/repository.rs [hunk 5]
```diff
@@ -221,9 +219,7 @@ impl Repository for Github {
                 } else {
                     "main".to_owned()
                 };
-                let (git_head, date_of_the_last_commit) = self.git_info().await?;
-                self.git_head = Some(git_head);
-                self.date_of_the_last_commit = Some(date_of_the_last_commit);
+                self.git_info().await?;
```
Comment (vincenzopalazzo) [inline coffee_github/src/repository.rs:222]: You do not need to have the method of the object, but the procedure can be static (without self) or just a macros, and this get_info method can return the information about the commit header and the commit time

File: coffee_github/src/repository.rs [hunk 6]
```diff
@@ -255,6 +251,9 @@ impl Repository for Github {
         }
         // pull the changes from the repository
         let status = fast_forward(&self.url.path_string, &self.branch)?;
+        // update the git information
+        // (git HEAD and date of the last commit)
+        self.git_info().await?;
```
Comment (vincenzopalazzo) [inline coffee_github/src/repository.rs:256]: I would like to see something like `let (commit, time) = get_repo_info!()`  and then so just call, `self.commit = commit`

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]
With the concept we are there but I think the implementation si mixing different concepts

File: coffee_core/src/coffee.rs (hunk 2 above)
Comment (vincenzopalazzo) [inline coffee_core/src/coffee.rs]: remove it and store inside the `CoffeeListRemote` the optional value

File: coffee_github/src/repository.rs [hunk 7]
```diff
@@ -36,6 +37,10 @@ pub struct Github {
     /// the name of the branch to be able to
     /// pull the changes from the correct branch
     branch: String,
+    /// the latest commit id of the repository
+    git_head: Option<String>,
+    /// the latest commit date of the repository
+    date_of_the_last_commit: Option<String>,
```
Comment (vincenzopalazzo) [inline coffee_github/src/repository.rs:43]: call it just `last_activity`

File: coffee_lib/src/repository.rs [hunk 8]
```diff
@@ -33,5 +33,14 @@ pub trait Repository: Any {
     /// return the url of the repository.
     fn url(&self) -> URL;
 
+    /// return the git head of the repository.
+    fn git_head(&self) -> Option<String>;
+
+    /// return the date of the last commit of the repository.
+    fn date_of_the_last_commit(&self) -> Option<String>;
```
Comment (vincenzopalazzo) [inline coffee_lib/src/repository.rs:40]: This is a bad design, you should remove the method for the `Repository` Trait. It is an implementation-specific information, so this should stay in `coffee_github`

File: coffee_lib/src/types/mod.rs [hunk 9]
```diff
@@ -109,6 +109,8 @@ pub mod response {
         pub local_name: String,
         pub url: String,
         pub plugins: Vec<Plugin>,
+        pub commit_id: String,
+        pub date: String,
```
Comment (vincenzopalazzo) [inline coffee_lib/src/types/mod.rs]: These need to be optional types

File: coffee_storage/src/model/repository.rs [hunk 10]
```diff
@@ -15,4 +15,6 @@ pub struct Repository {
     pub url: URL,
     pub plugins: Vec<Plugin>,
     pub branch: String,
+    pub git_head: Option<String>,
+    pub date_of_the_last_commit: Option<String>,
```
Comment (vincenzopalazzo) [inline coffee_storage/src/model/repository.rs:19]: move `date_of_the_last_commit` tp `last_activity`

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]
Needs some more works

File: coffee_core/src/coffee.rs (hunk 2 above)
Comment (tareknaser) [reply coffee_core/src/coffee.rs]: this is outdated
solved in 3b55f3e

File: coffee_lib/src/repository.rs (hunk 8 above)
Comment (tareknaser) [reply coffee_lib/src/repository.rs:40]: This is just the definition of these methods
The actual implementation is in `coffee_github/src/repository.rs`
We have to implement these methods here to be able to call them on variables on type `Repository` in `coffee_core/src/coffee.rs` 
For example in `list_remotes` method

Comment (vincenzopalazzo) [reply coffee_lib/src/repository.rs:40]: >This is just the definition of these methods

This is the problem. You require this method in all the implementations of this trait, and this is wrong. You should implement this only inside `coffee_github`.

>We have to implement these methods here to be able to call them on variables on type Repository in coffee_core/src/coffee.rs
For example, in list_remotes method

Why not call just `get_info` in these places?

Comment (tareknaser) [reply coffee_lib/src/repository.rs:40]: > Why not call just get_info in these places?
We definitely can

Comment (vincenzopalazzo) [reply coffee_lib/src/repository.rs]: yeah so we should only expose the get_info and then implement in the `coffee_github` the others two methods

File: coffee_cmd/src/coffee_term/command_show.rs [hunk 11]
```diff
@@ -54,15 +54,27 @@ pub fn show_remote_list(remote_list: Result<CoffeeRemote, CoffeeError>) -> Resul
...
 
     for repository in &repositories {
+        let commit_id = match &repository.commit_id {
+            Some(commit_id) => commit_id.to_owned(),
+            None => String::from(""),
+        };
+        let date = match &repository.date {
+            Some(date) => date.to_owned(),
+            None => String::from(""),
+        };
```
Comment (vincenzopalazzo) [inline coffee_cmd/src/coffee_term/command_show.rs:70]: You can use also `repository.date.unwrap_or_default()`

File: coffee_core/src/coffee.rs [hunk 12]
```diff
@@ -14,13 +14,14 @@ use serde::de::DeserializeOwned;
 use serde::{Deserialize, Serialize};
 use tokio::process::Command;
 
+use chrono::{TimeZone, Utc};
```
Comment (vincenzopalazzo) [inline coffee_core/src/coffee.rs:17]: this should be under the tokio import. We try to keep separate the our import and the library import.

File: coffee_core/src/coffee.rs [hunk 13]
```diff
@@ -14,13 +14,14 @@ use serde::de::DeserializeOwned;
...
 use coffee_github::repository::Github;
 use coffee_lib::errors::CoffeeError;
 use coffee_lib::plugin_manager::PluginManager;
 use coffee_lib::repository::Repository;
 use coffee_lib::types::response::*;
 use coffee_lib::url::URL;
-use coffee_lib::{error, sh};
+use coffee_lib::{commit_id, error, get_repo_info, sh};
```
Comment (vincenzopalazzo) [inline coffee_core/src/coffee.rs:24]: Now we should have the `use coffe_lib::prelude::*`

**Review by vincenzopalazzo — COMMENTED** [review]
LGTM

I have some comments but this looks good we can add them later

**Pull request merged**